Smart Pipes(reborn)

Ghilker opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
And we are back with smart pipes! Enjoy a new way of building pipes in an easy and more intuitive way.

[BASIC PIPING GIFs](https://imgur.com/a/BDueHKa)

How these new pipes work.
-Smart pipes autoconnect to nearby smart pipes
-They are now color coded, so they only connect to the same colored pipe, the GREY pipe is the wildcard and can connect to every other color, so be aware of this
-ALL components spawned by the RPD can be colored (from pumps to connectors, from pipes to manifolds), if you leave them GREY they can connect to every other color. Color adapters can be colored, but they'll still connect two pipes with different colors. BUILDABLE machines are GREY (thermomachines, cryo, HFR) so be aware of this
-Trying to go across another smart pipe will now build a bridge pipe automatically already colored of the color you choose, so you don't have to place it yourself anymore (is still available in the RPD tho)
-ALL binary components, layer manifolds, color adapters and bridge pipe can be put ONTOP of a smart pipe, but not on another of these. Smart pipes can't be placed on top of these pipes, so you have to build them first.
-Lcrossings can't be made anymore (sorry y'all i tryed, if someone have a way of doing them ping me on discord)
-REMEMBER you still have 5 layers to go, these rules apply to the same layer pipes, so if you do a crossing on different layers you won't see a bridge pipe appear.

Mappers should refer to [this comment](https://github.com/tgstation/tgstation/pull/58038#issuecomment-810986942) to learn how a few new rules on how to map with smart pipes
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Better pipe building
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Smart pipes, simple connection logic, similar to current power cable; place and they will automatically connect (also color dependable, grey is the wildcard that can connect to any colors)
add:How these new pipes work.
add:-Smart pipes autoconnect to nearby smart pipes
add:-They are now color coded, so they only connect to the same colored pipe, the GREY pipe is the wildcard and can connect to every other color, so be aware of this
add:-ALL components spawned by the RPD can be colored (from pumps to connectors, from pipes to manifolds), if you leave them GREY they can connect to every other color. Color adapters can be colored, but they'll still connect two pipes with different colors. BUILDABLE machines are GREY (thermomachines, cryo, HFR) so be aware of this
add:-Trying to go across another smart pipe will now build a bridge pipe automatically already colored of the color you choose, so you don't have to place it yourself anymore (is still available in the RPD tho)
add:-ALL binary components, layer manifolds, color adapters and bridge pipe can be put ONTOP of a smart pipe, but not on another of these. Smart pipes can't be placed on top of these pipes, so you have to build them first.
add:-Lcrossings can't be made anymore (sorry y'all i tryed, if someone have a way of doing them ping me on discord)
add:-REMEMBER you still have 5 layers to go, these rules apply to the same layer pipes, so if you do a crossing on different layers you won't see a bridge pipe appear.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Ghilker committed:
smart pipes

Ghilker committed:
oh yes

Sparkezel commented:
Could you show a gif of it in action, also what if I don't want to make an intersection?

Ghilker commented:
[Video here](https://vimeo.com/manage/videos/530304167)

Jhyrachy commented:
So there is a different pipenet for each color, even if they are on the same layer?

Ghilker committed:
stuff

Ghilker commented:
> So there is a different pipenet for each color, even if they are on the same layer?

yes, also the layers are going to be removed soon, so only colors will remain

Jhyrachy commented:
> yes, also the layers are going to be removed soon, so only colors will remain

Layer are good for glance value (how am I supposed to know if on the same tile there is a single pipe or there are 5?) and to have multiple 'not pipe objects' (pumps, connectors, ecc) on the same tile to save space.

Isn't better to just 'rename' layers to colors to keep visual clarity?

Ghilker commented:
> Layer are good for glance value (how am I supposed to know if on the same tile there is a single pipe or there are 5?) 

there will be indicators, like stripes on the pipe (if i understand how)

also i wish to keep layers, because people will have much more liberty of construction and ease in visual clarity

Jhyrachy commented:
My first tought, like an atmosian, was (if everything was stacked) to hide a plasma pipe under the distro one. Good luck finding it.

Atmos were never simple, but IMHO it should strive for visual clarity as much as possible.  
That's why I'm a huge fan of incresing atmos size as much as possible to reduce clutter and improve piping clarity, to allow even a less experienced player to quickly notice if something is 'out of place'

Ghilker commented:
i'm going to try a thing, i'll redo this, have only layers and no different color connection, colors will just be graphic thing

Jhyrachy commented:
color/layer aside

I really really like the idea of 'smartpipes'

And would be really nice if it was a settings available in the pipe dispenser, alongside the others.
Maybe just make that different colors do not interconnect.

Like this (sorry for my poor paint skills):
We start with a connector and a canister on a blue pipe
![immagine](https://user-images.githubusercontent.com/5394678/112850661-80856d00-90aa-11eb-82d3-a75a534c1f95.png)

If the color is still blue, we can add a side pipe, and it will merge with the existing pipenet
![immagine](https://user-images.githubusercontent.com/5394678/112850839-aad72a80-90aa-11eb-8762-c02d60f0135e.png)

But if the color is different (say green), it will not
![immagine](https://user-images.githubusercontent.com/5394678/112850930-c3474500-90aa-11eb-9893-975b895d0690.png)

Allowing us to just make them cross if needed
![immagine](https://user-images.githubusercontent.com/5394678/112851010-d9550580-90aa-11eb-8183-af84a74de15d.png)

This way we could make pipe fast, reducing cross contamination, increase flexibility and visual clarity (same color pipe does not automatically merge, you must manually select the specific pipe and use it)

Sorry if my explanation was not adequate

Ghilker commented:
Yeah i'm not going to remove layers, they aid a lot with visual clarity, i'll keep colors separated too because it helps with higher degree of customization of the loops.

Helianthus00 commented:
As long as this doesn't remove mapped directional pipes, I like this a lot and think it's a huge improvement. Are you planning on doing the same for heat exchange pipes?

FernandoJ8 commented:
Doesn't this make it so that you can't have separate pipenets of the same color run parallel to each other without joining together? This is a cool addition but it takes away precision, you should be able to turn it off for fine tuning, otherwise it's fairly limiting while only adding slight convenience.

CyoTheProtogen commented:
i don't like this, gonna fuck with my funny setups

FernandoJ8 commented:
Like for example, you can't do pipes crossing over perpendicular pipes like the ones in cooling loops cause they'd join to make a 4 way manifold. In the case of the space cooling loop it doesn't matter but for many things it does

FernandoJ8 commented:
I'm liking this less and less by the minute

FernandoJ8 commented:
and yes, you could make them different colours for that, but that means also adding a color adapter which cant' turn to the left or right and will also connect every other color

GuillaumePrata commented:
My main issue with it is that if you miss click and place a smart pipe you might connect 2 pipelines you didn't want and cleaning pipes is a pain in the ass.
Another issue is that if I miss click and make a manifold + side pipe, I can't safely remove that side pipe without spewing it's contents into the air or cleaning the full pipeline so again, more punishment for a simple miss click.
So hopefully we will have some more answers to make simple mistakes less punishing with smart pipes else it will just push new players away from Atmos since at least for me, it already looks more complicated to build things than the old system (Which I have a lot of experience of course so fucking biased...).

Ghilker commented:
> As long as this doesn't remove mapped directional pipes, I like this a lot and think it's a huge improvement. Are you planning on doing the same for heat exchange pipes?

I'm trying to still have directional pipes for map (tho i'll make them optional since the smart pipes are just "drag-n-draw"), but i can't promise anything rn, just that i'll try



> > you should be able to turn it off for fine tuning, otherwise it's fairly limiting while only adding slight convenience.
> 
> please make it optional or I will cry

no this is not going to be optional

> Like for example, you can't do pipes crossing over perpendicular pipes like the ones in cooling loops cause they'd join to make a 4 way manifold. In the case of the space cooling loop it doesn't matter but for many things it does
> 
> I'm liking this less and less by the minute
> 
> and yes, you could make them different colours for that, but that means also adding a color adapter which cant' turn to the left or right and will also connect every other color
> 
> I am a filthy comment flooder, I'm sorry

you can make crosspipes with different colors or use different layers to accomplish the same thing, this opens up a lot of new ways to create complex systems.

> My main issue with it is that if you miss click and place a smart pipe you might connect 2 pipelines you didn't want and cleaning pipes is a pain in the ass.
> Another issue is that if I miss click and make a manifold + side pipe, I can't safely remove that side pipe without spewing it's contents into the air or cleaning the full pipeline so again, more punishment for a simple miss click.
> So hopefully we will have some more answers to make simple mistakes less punishing with smart pipes else it will just push new players away from Atmos since at least for me, it already looks more complicated to build things than the old system (Which I have a lot of experience so I'm fucking biased...).

you'll learn to use this, don't worry is very easy

FernandoJ8 commented:
>
> you can make crosspipes with different colors or use different layers to accomplish the same thing, this opens up a lot of new ways to create complex systems.

There's still the issue that that forces you to add color adapters, which require a space that doesn't bend to the left or right per change / union of colours and will merge any pipes going through there. It limits how compact setups can be, which is an issue. Complexity isn't a problem ordinarily, on the contrary, the struggle is that you have a limited amount of space to work with, and this limits how much you can squeeze things further in some ways. 

The color idea I can get behind, but not having an option to set pipe directions manually seems like a bad idea

> you'll learn to use this, don't worry is very easy

That's kind of a dumb point to make when the point of smart pipes is making things easier, dealing with accidental pressurizations is already very annoying as  and is one of the hardest things to deal with. It isn't about learning, accidents happen, specially when it's something like misclicks and this makes it easier for them to happen

GuillaumePrata commented:
> you'll learn to use this, don't worry is very easy

I don't doubt that but mistakes will happen and the new system punishes it a lot.
You are building a pipenet with plasma, a random pushes you 1 tile and now you have a manifold + normal pipe and removing it will spew the plasma everywhere or worse, you contaminated 2 pipelines that were close.
Good luck cleaning a mistake like this around the HFR.
Guess we can always go with the get good and don't fuck up mentality...

Mothblocks commented:
Can you write a better changelog? This is a very breaking change, but it's not clear to anyone not already familiar with some niche stuff.

Compare to smart wiring's changelog:

![image](https://user-images.githubusercontent.com/35135081/112890128-55d7ec00-908b-11eb-8a6c-4a790f3ecd7f.png)

optimumtact commented:
it's difficult to get both colours & layers in a way that makes visual sense, so the clarity is hard to achieve.

I think it's possible, but you have to limit the maxn umber of pipes in a given tile to 0 -> layermax

Ghilker committed:
fef

Ghilker committed:
some map changes

Ghilker committed:
even more map

Ghilker committed:
this is only the start isn't it?

Ghilker committed:
pain

Ghilker committed:
wooo bridges

Ghilker committed:
fex

Zytolg commented:
I mentioned this in the discord, but I'd like to reiterate it here. My biggest concern about this PR was, and still is, how smartpipes in the map editors turned covered their turfs in a rainbowfied feverdream. This would be at its worst in atmos most likely, but aside from it looking less pristine, it would also add considerable effort into mapping atmos, as all mappers would be required to trace what goes where. Cables were able to avoid this since they are one color on one layer, however pipes are unfortunately a great many times more complex. It would be preferred if this visual nuance were to be somehow represented in the editors.

Ghilker committed:
better mapping

Ghilker committed:
merge master

Ghilker committed:
revert maps, revert color removal, all hail colors

Ghilker committed:
more mapping helpers

Ghilker commented:
To answer @Zytolg and all other mappers concers.
So the pipes will have directional sprites for mapping purpose but they'll behave as autoconnecting cables, so remember that this
<details> <summary> Map preview</summary>
<img src="https://user-images.githubusercontent.com/42839747/113134004-99b02a00-9220-11eb-8af4-7b3df5c4d5e5.png"> </details>
turns into this
<details> <summary> Game preview</summary>
<img src="https://user-images.githubusercontent.com/42839747/113134993-d4669200-9221-11eb-81c5-79089eeea0bd.png"> </details>

So how you do mapping now? Well is pretty much the same, just with a few new rules to respect.
First of all you'll have the new smart pipe list in the map editor
<details> <summary> Map preview</summary>
<img src="https://user-images.githubusercontent.com/42839747/113135233-2a3b3a00-9222-11eb-9c56-e27bee8e5c8c.PNG"> </details>

The piping will be the same, but it can be made in two ways actually
<details> <summary>Like this, as you can see pretty clean</summary>
<img src="https://user-images.githubusercontent.com/42839747/113135341-5060da00-9222-11eb-87d3-08ecda685dea.PNG"> </details>

<details> <summary>Or like this, more cluttered, but far far easier to make since you don't have to constantly change the direction of the sprite</summary>
<img src="https://user-images.githubusercontent.com/42839747/113135443-6d95a880-9222-11eb-9ce8-06feecbeeda4.PNG"> </details>

Another thing to remember now is that pipe crossing on the same layer is not possible in game (same color pipe will just connect)
<details> <summary>This is a no no</summary>
<img src="https://user-images.githubusercontent.com/42839747/113135664-a6ce1880-9222-11eb-992e-0384d5c32028.PNG"> </details>
So you'll have to use bridge pipes to make those 
<details> <summary>This is the proper way</summary>
<img src="https://user-images.githubusercontent.com/42839747/113135778-d3823000-9222-11eb-801e-7ca374e74ad1.PNG"> </details>
Those are colored to match the right pipe (also layer manifolds are colored)
Different layer pipes can still go across each other without issues
<details> <summary>Even if same color pipe</summary>
<img src="https://user-images.githubusercontent.com/42839747/113135962-147a4480-9223-11eb-8c1c-65e4d662c876.PNG"> </details>

Binary components will also act as bridges so feel free to use them as needed 
<details> <summary>Even if same color pipe</summary>
<img src="https://user-images.githubusercontent.com/42839747/113136115-3f649880-9223-11eb-84f8-8e1877b498be.PNG"> </details>

The color GREY (or the general pipes, or just the not colored ones) is the wildcard and can connect to all colors, so be aware of situations like this 
<details> <summary>Situation 1 </summary>
<img src="https://user-images.githubusercontent.com/42839747/113136248-7044cd80-9223-11eb-855b-981ea1789953.PNG"> </details>

<details> <summary>Situation 2 </summary>
<img src="https://user-images.githubusercontent.com/42839747/113136430-a4b88980-9223-11eb-87b1-af90d2db52e8.PNG"> </details>

Grey pipes will autoconnect to all other colors so if you need to do one of the various situations, just put the pipes on a different layer 
<details> <summary>Like this </summary>
<img src="https://user-images.githubusercontent.com/42839747/113136595-d893af00-9223-11eb-9790-3a328315b70b.PNG"> </details>

For any other question, ping me on discord @Ghilker

Ghilker commented:
Ok so, i changed how these pipes works, colors can't be overlapped anymore, so only one pipe per layer. 
Pipes will only connect if they are of the same color and on the same layer, grey pipes can be connected to all colors.

Layer manifolds are colored and can connect to only the choosen color (again except for grey)

Color adapters will connect pipe on different color and layer (i'm thinking of making them color only)

Pipes no longer are able to cross each other on the same layer (if on different color) due to it being abusable and the system very prone to errors when trying to handle multiple nodes in one, so i added bridge pipes that will allow the connection over a pipe easily. All binary devices are also able to be bridges (like pumps and valves).
Warning, to build bridges you have to first build the pipe underneath, you can't do the opposite.

Ghilker committed:
pipes can't be repainted with the RPD, color adapter now will connect different colored pipes and on one layer only

Mothblocks commented on code/__DEFINES/atmospherics.dm:
```diff
@@ -495,14 +499,29 @@ GLOBAL_LIST_INIT(pipe_paint_colors, sortList(list(
 		"dark" = rgb(69,69,69),
 		"green" = rgb(30,255,0),
-		"grey" = rgb(255,255,255),
+		"grey" = rgb(254,254,254),
```
These *have* to become defines. They're referenced more or less everywhere, meaning wanting to change the color is a way bigger pain in the ass than it has any right to be, like how this specific value of 254, 254, 254 crops up a lot.

Mothblocks commented on code/__DEFINES/atmospherics.dm:
```diff
@@ -495,14 +499,29 @@ GLOBAL_LIST_INIT(pipe_paint_colors, sortList(list(
 )))
 
+GLOBAL_LIST_INIT(pipe_colors_ordered, sortList(list(
+		rgb(130,43,255) = -6,
+		rgb(0,0,255) = -5,
+		rgb(178,100,56) = -4,
+		rgb(0,255,249) = -3,
+		rgb(69,69,69) = -2,
+		rgb(30,255,0) = -1,
+		rgb(254,254,254) = 0,
+		rgb(255,129,25) = 1,
+		rgb(128,0,182) = 2,
+		rgb(255,0,0) = 3,
+		rgb(64,0,128) = 4,
+		rgb(255,198,0) = 5
```
Only one extra layer of indentation, not several.

Mothblocks commented on code/game/objects/items/RPD.dm:
```diff
@@ -471,7 +452,7 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 						P.add_fingerprint(usr)
 						P.setPipingLayer(piping_layer)
-						if(findtext("[queued_p_type]", "/obj/machinery/atmospherics/pipe") && !findtext("[queued_p_type]", "layer_manifold"))
+						if(findtext("[queued_p_type]", "/obj/machinery/atmospherics") && !findtext("[queued_p_type]", "color_adapter"))
```
```suggestion
						if(ispath(queued_p_type, /obj/machinery/atmospherics) && !ispath(queued_p_type, /obj/machinery/atmospherics/pipe/color_adapter))
```

Mothblocks commented on code/modules/atmospherics/machinery/atmosmachinery.dm:
```diff
@@ -228,7 +233,8 @@
 	if(isnull(given_layer))
 		given_layer = piping_layer
-	if((target.piping_layer == given_layer) || (target.pipe_flags & PIPING_ALL_LAYER))
+	if(	((target.piping_layer == given_layer) || (target.pipe_flags & PIPING_ALL_LAYER)) && target.loc != loc && \
+		(target.pipe_color == pipe_color || target.pipe_flags & PIPING_ALL_COLORS || pipe_flags & PIPING_ALL_COLORS || target.pipe_color == rgb(254,254,254) || pipe_color == rgb(254,254,254)))
```
```suggestion
	if(
		((target.piping_layer == given_layer) || (target.pipe_flags & PIPING_ALL_LAYER)) && target.loc != loc \
		&& (target.pipe_color == pipe_color || target.pipe_flags & PIPING_ALL_COLORS || pipe_flags & PIPING_ALL_COLORS || target.pipe_color == rgb(254,254,254) || pipe_color == rgb(254,254,254))
	)
```

If you're gonna do multi-lining, then it's gotta be correct, no random indentations.

The `&&` on its own line makes it easier to know what the next operation is without going to the end of the previous line.

Mothblocks commented on code/modules/atmospherics/machinery/pipes/bridge_pipe.dm:
```diff
@@ -0,0 +1,46 @@
+
+	var/static/list/mutable_appearance/center_cache = list()
+
+/obj/machinery/atmospherics/pipe/bridge_pipe/New()
```
Not `Initialize`?

Ghilker replied:
Added a comment explaining why

ninjanomnom replied:
Deal with it in another way even if it ends up being that you update appearance after init, there's no reason nowadays to be using New() except on datums.

Ghilker replied:
moved to init next commit

Mothblocks commented on code/modules/atmospherics/machinery/pipes/color_adapter.dm:
```diff
@@ -0,0 +1,60 @@
+
+	var/static/list/mutable_appearance/center_cache = list()
+
+/obj/machinery/atmospherics/pipe/color_adapter/New()
```
Same here.

Ghilker replied:
same

ninjanomnom replied:
same

Ghilker replied:
same

Mothblocks commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,114 @@
+	var/static/list/mutable_appearance/center_cache = list()
+	var/mutable_appearance/pipe_appearance
+
+/obj/machinery/atmospherics/pipe/smart/New()
```
`/Initialize`

Ghilker replied:
same

ninjanomnom replied:
same

Ghilker replied:
same

Mothblocks commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,114 @@
+	connections = list(dir2text(NORTH) = FALSE, dir2text(SOUTH) = FALSE , dir2text(EAST) = FALSE , dir2text(WEST) = FALSE)
+	var/list/valid_connectors = typecacheof(/obj/machinery/atmospherics)
+	for(var/direction in connections)
+		var/turf/T = get_step(src,  text2dir(direction))
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Also, fix the random spaces.

Mothblocks commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,114 @@
+				if(connections[direction] != TRUE)
+					continue
+				center = mutable_appearance('icons/obj/atmospherics/pipes/simple.dmi', "pipe00-3")
+				src.dir = text2dir(direction)
```
```suggestion
				dir = text2dir(direction)
```

Mothblocks commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,114 @@
+			for(var/direction in connections)
+				if(connections[direction] == FALSE)
+					center = mutable_appearance('icons/obj/atmospherics/pipes/manifold.dmi', "manifold_center")
+					src.dir = text2dir(direction)
```
```suggestion
					dir = text2dir(direction)
```

Mothblocks commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,114 @@
+					src.dir = text2dir(dir2text(text2dir(direction)+text2dir(direction2)))
+		if(3)
+			for(var/direction in connections)
+				if(connections[direction] == FALSE)
```
You do `== FALSE` here, yet `!= TRUE` in other places. I say make them all a consistent `if (!connections[direction])`.

Mothblocks commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,114 @@
+					src.dir = text2dir(direction)
+		if(4)
+			center = mutable_appearance('icons/obj/atmospherics/pipes/manifold.dmi', "manifold4w_center")
+			src.dir = dir2text(NORTH)
```
```suggestion
			dir = dir2text(NORTH)
```

Mothblocks commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,114 @@
+/obj/machinery/atmospherics/pipe/smart/proc/check_connections()
+	var/mutable_appearance/center
+	connection_num = 0
+	connections = list(dir2text(NORTH) = FALSE, dir2text(SOUTH) = FALSE , dir2text(EAST) = FALSE , dir2text(WEST) = FALSE)
```
I think a bitfield for this would be better than an assoc list, since all the directionals are bitflags anyway. Would likely make some other stuff nicer.

Ghilker replied:
I tried to make this a bitfield, but the directions are weird and i cannot manage to make it work properly, mostly because the pipes need a very precise direction check and with the bitfield it seems to be dependable on the order you build the pipes

ninjanomnom replied:
The moth is right here, we can figure out the issue more in coderbus but if the issue is just in being able to iterate over the direction, that's what `GLOB.cardinals` is for

Or the slightly too clever
```
for(var/i in 0 to 3)
  var/dir = 1 << i
```

Ghilker replied:
done, used flag instead of the dir2text mess

Mothblocks commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,114 @@
+	PIPING_LAYER_DOUBLE_SHIFT(center, piping_layer)
+	center_cache["[piping_layer]"] = center
+	pipe_appearance = center
+	. +=center
```
```suggestion
	. += center
```

Consistency.

ninjanomnom replied:
Don't resolve things without either changing the code or saying why in a comment

Ghilker replied:
oof, i thought i fixed that, apparently not

FernandoJ8 commented:
> Pipes no longer are able to cross each other on the same layer (if on different color) due to it being abusable and the system very prone to errors when trying to handle multiple nodes in one, so i added bridge pipes that will allow the connection over a pipe easily. All binary devices are also able to be bridges (like pumps and valves).
> Warning, to build bridges you have to first build the pipe underneath, you can't do the opposite.

Are bridges going to split pipenets into two or will they act like a straight pipe?

Ghilker commented:
> Are bridges going to split pipenets into two or will they act like a straight pipe?

they act like straight pipes

Ghilker committed:
changes

Ghilker commented:
best to leave it as draft since it need map changes

Ghilker committed:
AHHHHHHHHHHH

Ghilker committed:
AHHHH

Ghilker committed:
*scream*

Ghilker committed:
last one? *more screams*

Ghilker committed:
manifold *screams*

Ghilker committed:
PLEASEEE

Ghilker committed:
all stations done

Ghilker committed:
remove

Ghilker committed:
poke

Ghilker committed:
major fixes

Ghilker committed:
map fix

ninjanomnom commented on code/modules/atmospherics/machinery/atmosmachinery.dm:
```diff
@@ -228,7 +233,10 @@
 	if(isnull(given_layer))
 		given_layer = piping_layer
-	if((target.piping_layer == given_layer) || (target.pipe_flags & PIPING_ALL_LAYER))
+	if(
+		((target.piping_layer == given_layer) || (target.pipe_flags & PIPING_ALL_LAYER)) && target.loc != loc \
+		&& (target.pipe_color == pipe_color || target.pipe_flags & PIPING_ALL_COLORS || pipe_flags & PIPING_ALL_COLORS || target.pipe_color == COLOR_VERY_LIGHT_GRAY || pipe_color == COLOR_VERY_LIGHT_GRAY)
+	)
```
make this into a proc so it can be written more sane

Ghilker replied:
made the if string into 2 procs (one for layer and one for color)

ninjanomnom commented on code/modules/atmospherics/machinery/other/meter.dm:
```diff
@@ -3,7 +3,7 @@
 	icon = 'icons/obj/atmospherics/pipes/meter.dmi'
 	icon_state = "meterX"
-	layer = GAS_PUMP_LAYER
+	layer = 2.55
```
Need to be in relation to a define, did you mean to make this number just over the gas pump layer? If so this sort of thing is decent
```suggestion
	layer = GAS_PUMP_LAYER + 0.01
```
Otherwise you probably want a new define in here
https://github.com/tgstation/tgstation/blob/f4a280c4b67218b619c7a256a214fe0f8b4c1f61/code/__DEFINES/layers.dm#L50-L55

Ghilker replied:
added a define

ninjanomnom commented on code/modules/atmospherics/machinery/pipes/bridge_pipe.dm:
```diff
@@ -0,0 +1,49 @@
+		center_cache["[piping_layer]"] = center
+	. += center
+
+	layer = 2.55 //to stay above all sorts of pipes
```
Again, needs to be in relation to a define

Ghilker replied:
done

ninjanomnom commented on code/modules/atmospherics/machinery/pipes/bridge_pipe.dm:
```diff
@@ -0,0 +1,49 @@
+
+	//Add non-broken pieces
+	for(var/i in 1 to device_type)
+		if(nodes[i])
```
Avoid nesting by inverting this check and doing `continue`

Ghilker replied:
done

ninjanomnom commented on code/modules/atmospherics/machinery/pipes/color_adapter.dm:
```diff
@@ -0,0 +1,64 @@
+
+	//Add non-broken pieces
+	for(var/i in 1 to device_type)
+		if(nodes[i])
```
Same comment here regarding avoiding nesting

Ghilker replied:
ye

ninjanomnom commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,115 @@
+	for(var/direction in connections)
+		var/turf/turf = get_step(src, text2dir(direction))
+		if(!turf)
+			return
```
I assume you didnt want to exit this entire proc just because the edge of the map is one tile over

Ghilker replied:
yes changed to a continue

LemonInTheDark commented on code/modules/atmospherics/machinery/components/components_base.dm:
```diff
@@ -100,7 +101,7 @@
 
 /obj/machinery/atmospherics/components/on_construction()
-	..()
+	. = ..()
```
does this change behavior at all? or is it just a consistency thing

Ghilker replied:
i changed it to test if it was this that wasn't making a thing work, but then i left it for consistency

LemonInTheDark commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,115 @@
+		var/turf/turf = get_step(src, text2dir(direction))
+		if(!turf)
+			return
+		for(var/machine_type in turf.contents)
```
this is fucking with my head, why are you using a typecache for this?

Ghilker replied:
to be able to quickly search... what i should use instead?

Ghilker replied:
i changed the system, now it will use just a for loop with a typecheck

LemonInTheDark commented on code/modules/atmospherics/machinery/pipes/smart.dm:
```diff
@@ -0,0 +1,115 @@
+				if(!connections[direction])
+					continue
+				//Detects straight pipes connected from east to west , north to south etc.
+				if(connections[dir2text(angle2dir(dir2angle(text2dir(direction))+180))] == TRUE)
```
this is really silly

Ghilker replied:
i got it removed (yes it was dumb)

LemonInTheDark commented on code/modules/atmospherics/machinery/atmosmachinery.dm:
```diff
@@ -198,9 +198,14 @@
 	for(var/obj/machinery/atmospherics/target in get_step(src, direction))
 		if(target.initialize_directions & get_dir(target,src))
+			if(connect_all)
+				if(target.pipe_color != GLOB.pipe_paint_colors[color_piping] || target.piping_layer != prompted_layer)
```
why does connect all exist in the first place? shouldn't this logic be a part of connection_check and it's subprocs?

Ghilker replied:
ah it was a remnant of old code while i was working, i'll remove it

ninjanomnom commented on code/__DEFINES/colors.dm:
```diff
@@ -167,3 +169,34 @@
 /// Barely visible cyan-ish hue, as the doctor prescribed. rgb(240, 250, 250)
 #define LIGHT_COLOR_HALOGEN    "#F0FAFA"
+
+///Color of pipes, moved here from atmospherics.dm due to the use of defines that are unreachable there
+GLOBAL_LIST_INIT(pipe_paint_colors, sortList(list(
+	"amethyst" = COLOR_AMETHYST, //supplymain
+	"blue" = COLOR_BLUE,
+	"brown" = COLOR_BROWN,
+	"cyan" = COLOR_CYAN,
+	"dark" = COLOR_DARK,
+	"green" = COLOR_VIBRANT_LIME,
+	"grey" = COLOR_VERY_LIGHT_GRAY,
+	"orange" = COLOR_TAN_ORANGE,
+	"purple" = COLOR_PURPLE,
+	"red" = COLOR_RED,
+	"violet" = COLOR_STRONG_VIOLET,
+	"yellow" = COLOR_YELLOW
+)))
+
+GLOBAL_LIST_INIT(pipe_colors_ordered, sortList(list(
+	COLOR_AMETHYST = -6,
+	COLOR_BLUE = -5,
+	COLOR_BROWN = -4,
+	COLOR_CYAN = -3,
+	COLOR_DARK = -2,
+	COLOR_VIBRANT_LIME = -1,
+	COLOR_VERY_LIGHT_GRAY = 0,
+	COLOR_TAN_ORANGE = 1,
+	COLOR_PURPLE = 2,
+	COLOR_RED = 3,
+	COLOR_STRONG_VIOLET = 4,
+	COLOR_YELLOW = 5
+)))
```
rather than renaming the colors file, these lists should be in __HELPERS

Ghilker replied:
moved to piping_colors_list.dm inside _HELPERS

Ghilker commented on code/__DEFINES/colors.dm:
```diff
@@ -8,15 +8,16 @@
 #define COLOR_DARKMODE_TEXT "#a4bad6"
 
 #define COLOR_WHITE "#FFFFFF"
-#define COLOR_VERY_LIGHT_GRAY "#EEEEEE"
```
because the RPD set the colors as undercase "#eeeeee" while the mapped ones are in uppercase since they use these defines. This caused that the mapped pipes couldn't connect at all to the RPD ones. 
I tried to just change the RPD colors in the .js to be all uppercases, but they'll still get set to undercases, so i had to change them on the defines

ninjanomnom replied:
When comparing hex rgb values you should be doing lowertext() to both sides wherever the comparison is happening if it's possible to have the wrong case come in as input

Ghilker replied:
So I revert the color.dm change and put lowertext() to the checks?

Mothblocks replied:
preferably yeah

Ghilker replied:
done, next commit

Ghilker commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -59,6 +61,7 @@ Buildable meters
 	add_atom_colour(make_from.color, FIXED_COLOUR_PRIORITY)
 	pipe_type = make_from.type
 	paintable = make_from.paintable
+	pipe_color = make_from.pipe_color
```
for some reason the color (atom one) was different from pipe_color, so the components didn't work. I tested before changing this line and components didn't retain their colors (it was set to null after wrenching them back) and it worked after just changing this

Ghilker commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -141,16 +144,21 @@ Buildable meters
 			to_chat(user, "<span class='warning'>Something is hogging the tile!</span>")
 			return TRUE
+
+		if(flags & PIPING_BRIDGE && !(M.pipe_flags & PIPING_BRIDGE)) //continue if we are placing a bridge pipe over a normal pipe only (prevent duplicates)
```
i tried something, but it made it worse (pipes could again overlap over each other but for some reasons didn't connect)
if you find some better way please lemme know

Ghilker replied:
i found a way, just place a bridge pipe automatically lol thanks to @ninjanomnom for the idea

Ghilker commented on code/game/objects/items/RPD.dm:
```diff
@@ -404,7 +403,7 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 	if(!make_pipe_whitelist)
 		make_pipe_whitelist = typecacheof(list(/obj/structure/lattice, /obj/structure/girder, /obj/item/pipe, /obj/structure/window, /obj/structure/grille))
-	if(istype(attack_target, /obj/machinery/atmospherics) && (mode & BUILD_MODE && !(mode & PAINT_MODE))) //Reduces pixelhunt when coloring is off.
+	if(istype(attack_target, /obj/machinery/atmospherics) && mode & BUILD_MODE) //Reduces pixelhunt when coloring is off.
```
yeeted

Ghilker commented on code/game/objects/items/RPD.dm:
```diff
@@ -471,7 +452,7 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 						P.add_fingerprint(usr)
 						P.setPipingLayer(piping_layer)
-						if(findtext("[queued_p_type]", "/obj/machinery/atmospherics/pipe") && !findtext("[queued_p_type]", "layer_manifold"))
+						if(ispath(queued_p_type, /obj/machinery/atmospherics) && !ispath(queued_p_type, /obj/machinery/atmospherics/pipe/color_adapter))
```
all thanks to moth

Ghilker committed:
main code improvements

Ghilker committed:
icebox fix

Ghilker committed:
woops double pipe

Ghilker committed:
multiz fix, tramstation fix, static list derps

Ghilker committed:
fex

Ghilker committed:
auto bridge pipe, layermanifolds and color adapters can be placed across pipes, better smart pipe icon logic, hopefully better performances

Ghilker committed:
manifolds

Mothblocks commented on code/__HELPERS/piping_colors_lists.dm:
```diff
@@ -0,0 +1,31 @@
+///All colors available to pipes and atmos components
+GLOBAL_LIST_INIT(pipe_paint_colors, sortList(list(
+	"amethyst" = COLOR_AMETHYST, //supplymain
```
Why does only this one have a comment? What does it mean?

Ghilker replied:
Nothing is a remnant of the old code, probably amethyst was used as supply pipe.

Want me to remove it?

Ghilker replied:
done, next commit

Mothblocks commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -137,20 +149,39 @@ Buildable meters
 	var/obj/machinery/atmospherics/fakeA = pipe_type
 	var/flags = initial(fakeA.pipe_flags)
+	var/pipe_count = 0
+	for(var/obj/machinery/atmospherics/M in loc)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Ghilker replied:
done, next commit

Mothblocks commented on code/modules/atmospherics/machinery/atmosmachinery.dm:
```diff
@@ -407,6 +436,12 @@
 
+/obj/machinery/atmospherics/proc/pipe_overlay(set_icon, set_state, direction, color = COLOR_VERY_LIGHT_GRAY, piping_layer = 3, set_layer = PIPE_VISIBLE_LEVEL)
+	var/image/pipe_overlay
+	pipe_overlay = . = image(icon = set_icon, icon_state = set_state, layer = set_layer, dir = direction)
```
```suggestion
	pipe_overlay = image(icon = set_icon, icon_state = set_state, layer = set_layer, dir = direction)
```

Then `return pipe_overlay` at the end. This is needlessly confusing.

Ghilker replied:
done, next commit

Mothblocks commented on tgui/packages/tgui/interfaces/RapidPipeDispenser.js:
```diff
@@ -22,17 +22,17 @@ const ICON_BY_CATEGORY_NAME = {
-  red: '#ff3333',
-  violet: '#6e00f6',
-  yellow: '#ffce26',
+  amethyst: '#822bff',
+  blue: '#0000ff',
+  brown: '#ba9f6d',
+  cyan: '#00ffff',
+  dark: '#454545',
+  green: '#00ff00',
+  orange: '#ff7b00',
+  purple: '#800080',
+  red: '#ff0000',
+  violet: '#6927c5',
+  yellow: '#ffff00',
```
Might be too late, but I wish these were in `ui_static_data`, or some other code-sourced place.

LemonInTheDark replied:
It's never too late

Ghilker replied:
Why you want me to do more tgui 
![cat](https://user-images.githubusercontent.com/42839747/113552382-5ece3d80-95f6-11eb-8d62-b4bc492ffd7e.jpg)

Ghilker replied:
done, next commit

Ghilker committed:
merge master

Ghilker committed:
requested changes, better pipe connectors handling

Ghilker committed:
more map fix

Ghilker committed:
fix

ninjanomnom commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -137,23 +150,43 @@ Buildable meters
 			return TRUE
-		if((M.piping_layer != piping_layer) && !((M.pipe_flags | flags) & PIPING_ALL_LAYER)) //don't continue if either pipe goes across all layers
+
+		if(pipe_count == 1 && istype(machine, /obj/machinery/atmospherics/pipe/smart) && pipe_type == /obj/machinery/atmospherics/pipe/smart && lowertext(machine.pipe_color) != lowertext(pipe_color) && machine.connection_num < 3)
```
Why are you doing a direct type comparison here? You really should not be doing this sort of thing.

Ghilker replied:
Changed to ispath(pipe_type, path)

Unless I misunderstood something

Ghilker committed:
Update construction.dm

ninjanomnom commented:
>2 hour long map diff bot

The other checks have all passed and map diff seems to be stuck having completed but not exiting.

LemonInTheDark commented:
I was reviewing that

AkiraSipes commented:
It sucks ass, every time this is merged we cry a little more.

optimumtact commented:
> 
> 
> It sucks ass, every time this is merged we cry a little more.

no problem it will only be merged once

Ghilker commented:
[BASIC PIPING GIFs](https://imgur.com/a/BDueHKa)
